File: app/src/main/java/com/zoomia/app/MainActivity.java
package com.zoomia.app;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.InputType;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class MainActivity extends Activity {
    private EditText endpointInput;
    private EditText modelInput;
    private EditText keyInput;
    private EditText promptInput;
    private TextView conversation;
    private Button sendButton;
    private SharedPreferences prefs;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        prefs = getSharedPreferences("zoom_ia_settings", Context.MODE_PRIVATE);
        setContentView(createUi());
    }

    private View createUi() {
        ScrollView scroll = new ScrollView(this);
        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setPadding(32, 32, 32, 48);
        scroll.addView(root);

        TextView title = new TextView(this);
        title.setText("Zoom IA");
        title.setTextSize(30f);
        root.addView(title);

        TextView subtitle = new TextView(this);
        subtitle.setText("Sua IA Android editavel");
        subtitle.setTextSize(16f);
        root.addView(subtitle);

        addSpace(root, 20);
        addSection(root, "Configuracao da IA");

        endpointInput = addInput(root, "Endpoint completo, ex.: https://seu-servidor/v1/chat/completions");
        endpointInput.setText(prefs.getString("endpoint", ""));

        modelInput = addInput(root, "Modelo");
        modelInput.setText(prefs.getString("model", ""));

        keyInput = addInput(root, "Chave da API (opcional)");
        keyInput.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
        keyInput.setText(prefs.getString("api_key", ""));

        Button save = new Button(this);
        save.setText("Salvar configuracao");
        save.setOnClickListener(v -> {
            prefs.edit()
                    .putString("endpoint", endpointInput.getText().toString().trim())
                    .putString("model", modelInput.getText().toString().trim())
                    .putString("api_key", keyInput.getText().toString().trim())
                    .apply();
            Toast.makeText(this, "Configuracao salva", Toast.LENGTH_SHORT).show();
        });
        root.addView(save);

        addSpace(root, 20);
        addSection(root, "Ferramentas");
        addTool(root, "Chat IA", "Conecta a uma API compativel com /v1/chat/completions.");
        addTool(root, "Gerar imagem", "Pronto para conectar a um gerador de imagens.");
        addTool(root, "Remover objeto", "Pronto para conectar a inpainting e mascara.");
        addTool(root, "Trocar/adicionar roupa", "Pronto para conectar a virtual try-on.");
        addTool(root, "Editar video", "Pronto para conectar a um modelo de video.");

        addSpace(root, 20);
        addSection(root, "Conversa");
        conversation = new TextView(this);
        conversation.setText("Configure sua API acima e envie uma mensagem.\n");
        conversation.setTextSize(16f);
        root.addView(conversation);

        promptInput = addInput(root, "Digite um pedido para sua IA");
        promptInput.setMinLines(2);

        sendButton = new Button(this);
        sendButton.setText("Enviar");
        sendButton.setOnClickListener(v -> sendMessage());
        root.addView(sendButton);

        return scroll;
    }

    private void sendMessage() {
        String endpoint = endpointInput.getText().toString().trim();
        String model = modelInput.getText().toString().trim();
        String key = keyInput.getText().toString().trim();
        String prompt = promptInput.getText().toString().trim();

        if (prompt.isEmpty()) return;
        conversation.append("\nVoce: " + prompt + "\n");
        promptInput.setText("");

        if (endpoint.isEmpty() || model.isEmpty()) {
            conversation.append("Zoom IA: preencha o endpoint e o modelo primeiro.\n");
            return;
        }

        sendButton.setEnabled(false);
        sendButton.setText("Enviando...");

        executor.execute(() -> {
            String answer = callChatApi(endpoint, model, key, prompt);
            runOnUiThread(() -> {
                conversation.append("Zoom IA: " + answer + "\n");
                sendButton.setEnabled(true);
                sendButton.setText("Enviar");
            });
        });
    }

    private String callChatApi(String endpoint, String model, String key, String prompt) {
        HttpURLConnection connection = null;
        try {
            connection = (HttpURLConnection) new URL(endpoint).openConnection();
            connection.setRequestMethod("POST");
            connection.setConnectTimeout(30000);
            connection.setReadTimeout(60000);
            connection.setDoOutput(true);
            connection.setRequestProperty("Content-Type", "application/json");
            if (!key.isEmpty()) {
                connection.setRequestProperty("Authorization", "Bearer " + key);
            }

            JSONObject message = new JSONObject();
            message.put("role", "user");
            message.put("content", prompt);

            JSONArray messages = new JSONArray();
            messages.put(message);

            JSONObject payload = new JSONObject();
            payload.put("model", model);
            payload.put("messages", messages);

            byte[] data = payload.toString().getBytes(StandardCharsets.UTF_8);
            try (OutputStream os = connection.getOutputStream()) {
                os.write(data);
            }

            int code = connection.getResponseCode();
            InputStream stream = code >= 200 && code < 300
                    ? connection.getInputStream()
                    : connection.getErrorStream();
            String body = readAll(stream);

            if (code < 200 || code >= 300) {
                return "Erro HTTP " + code + ": " + trim(body, 500);
            }

            JSONObject root = new JSONObject(body);
            JSONArray choices = root.optJSONArray("choices");
            if (choices == null || choices.length() == 0) {
                return "A API respondeu, mas nao retornou choices[0].message.content.";
            }
            JSONObject messageObject = choices.optJSONObject(0).optJSONObject("message");
            if (messageObject == null) {
                return "Resposta recebida, mas sem o campo message.";
            }
            String content = messageObject.optString("content", "").trim();
            return content.isEmpty() ? "Resposta vazia da API." : content;
        } catch (Exception e) {
            return "Falha ao conectar: " + (e.getMessage() == null ? e.getClass().getSimpleName() : e.getMessage());
        } finally {
            if (connection != null) connection.disconnect();
        }
    }

    private String readAll(InputStream stream) throws Exception {
        if (stream == null) return "";
        StringBuilder sb = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8))) {
            String line;
            while ((line = reader.readLine()) != null) sb.append(line);
        }
        return sb.toString();
    }

    private String trim(String value, int max) {
        if (value == null) return "";
        return value.length() <= max ? value : value.substring(0, max);
    }

    private EditText addInput(LinearLayout root, String hint) {
        EditText input = new EditText(this);
        input.setHint(hint);
        input.setSingleLine(false);
        root.addView(input, new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT));
        return input;
    }

    private void addSection(LinearLayout root, String text) {
        TextView view = new TextView(this);
        view.setText(text);
        view.setTextSize(21f);
        root.addView(view);
    }

    private void addTool(LinearLayout root, String name, String description) {
        TextView view = new TextView(this);
        view.setText("\n" + name + "\n" + description);
        view.setTextSize(16f);
        view.setPadding(12, 10, 12, 10);
        root.addView(view);
    }

    private void addSpace(LinearLayout root, int height) {
        View space = new View(this);
        root.addView(space, new LinearLayout.LayoutParams(1, height));
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        executor.shutdownNow();
    }
}

File: app/src/main/java/com/zoomia/app/MainActivity.kt
package com.zoomia.app

import android.content.Context
import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            MaterialTheme {
                Surface(modifier = Modifier.fillMaxSize()) {
                    ZoomIAApp()
                }
            }
        }
    }
}

data class ChatLine(val author: String, val text: String)

data class ToolItem(val title: String, val description: String)

@Composable
fun ZoomIAApp() {
    val context = LocalContext.current
    val prefs = remember {
        context.getSharedPreferences("zoom_ia_settings", Context.MODE_PRIVATE)
    }
    val scope = rememberCoroutineScope()

    var endpoint by remember {
        mutableStateOf(prefs.getString("endpoint", "") ?: "")
    }
    var model by remember {
        mutableStateOf(prefs.getString("model", "") ?: "")
    }
    var apiKey by remember {
        mutableStateOf(prefs.getString("api_key", "") ?: "")
    }
    var prompt by remember { mutableStateOf("") }
    var isSending by remember { mutableStateOf(false) }
    var status by remember { mutableStateOf("Configure sua API para começar.") }

    val messages = remember { mutableStateListOf<ChatLine>() }

    val tools = listOf(
        ToolItem("Chat IA", "Funciona com uma API de chat compatível com OpenAI."),
        ToolItem("Gerar imagem", "Módulo preparado para conectar ao seu gerador de imagens."),
        ToolItem("Remover objeto", "Módulo preparado para inpainting/máscara."),
        ToolItem("Trocar roupa", "Módulo preparado para virtual try-on."),
        ToolItem("Editar vídeo", "Módulo preparado para geração e edição de vídeo.")
    )

    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(16.dp)
    ) {
        Text("Zoom IA", style = MaterialTheme.typography.headlineLarge)
        Text("Sua IA Android editável")
        Spacer(Modifier.height(12.dp))

        LazyColumn(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(10.dp)
        ) {
            item {
                Card(modifier = Modifier.fillMaxWidth()) {
                    Column(modifier = Modifier.padding(14.dp)) {
                        Text("Configuração da IA", style = MaterialTheme.typography.titleMedium)
                        Spacer(Modifier.height(8.dp))
                        OutlinedTextField(
                            modifier = Modifier.fillMaxWidth(),
                            value = endpoint,
                            onValueChange = { endpoint = it },
                            label = { Text("Endpoint /v1/chat/completions") },
                            singleLine = true
                        )
                        Spacer(Modifier.height(8.dp))
                        OutlinedTextField(
                            modifier = Modifier.fillMaxWidth(),
                            value = model,
                            onValueChange = { model = it },
                            label = { Text("Modelo") },
                            singleLine = true
                        )
                        Spacer(Modifier.height(8.dp))
                        OutlinedTextField(
                            modifier = Modifier.fillMaxWidth(),
                            value = apiKey,
                            onValueChange = { apiKey = it },
                            label = { Text("Chave da API") },
                            singleLine = true,
                            visualTransformation = PasswordVisualTransformation()
                        )
                        Spacer(Modifier.height(8.dp))
                        Button(onClick = {
                            prefs.edit()
                                .putString("endpoint", endpoint.trim())
                                .putString("model", model.trim())
                                .putString("api_key", apiKey.trim())
                                .apply()
                            status = "Configuração salva neste aparelho."
                        }) {
                            Text("Salvar configuração")
                        }
                        Spacer(Modifier.height(6.dp))
                        Text(status, style = MaterialTheme.typography.bodySmall)
                    }
                }
            }

            item {
                Text("Ferramentas", style = MaterialTheme.typography.titleMedium)
            }

            items(tools) { tool ->
                Card(modifier = Modifier.fillMaxWidth()) {
                    Column(modifier = Modifier.padding(14.dp)) {
                        Text(tool.title, style = MaterialTheme.typography.titleMedium)
                        Text(tool.description)
                    }
                }
            }

            if (messages.isNotEmpty()) {
                item {
                    HorizontalDivider()
                    Spacer(Modifier.height(8.dp))
                    Text("Conversa", style = MaterialTheme.typography.titleMedium)
                }
                items(messages) { message ->
                    Text("${message.author}: ${message.text}")
                }
            }
        }

        OutlinedTextField(
            modifier = Modifier.fillMaxWidth(),
            value = prompt,
            onValueChange = { prompt = it },
            label = { Text("Digite um pedido") },
            enabled = !isSending
        )
        Spacer(Modifier.height(8.dp))
        Row(modifier = Modifier.fillMaxWidth()) {
            Button(
                modifier = Modifier.fillMaxWidth(),
                enabled = !isSending,
                onClick = {
                    val userPrompt = prompt.trim()
                    val currentEndpoint = endpoint.trim()
                    val currentModel = model.trim()
                    val currentKey = apiKey.trim()

                    if (userPrompt.isBlank()) return@Button
                    messages.add(ChatLine("Você", userPrompt))
                    prompt = ""

                    if (currentEndpoint.isBlank() || currentModel.isBlank()) {
                        messages.add(
                            ChatLine(
                                "Zoom IA",
                                "Preencha o endpoint e o modelo na configuração acima."
                            )
                        )
                        return@Button
                    }

                    isSending = true
                    scope.launch {
                        val answer = withContext(Dispatchers.IO) {
                            callChatApi(
                                endpoint = currentEndpoint,
                                model = currentModel,
                                apiKey = currentKey,
                                prompt = userPrompt
                            )
                        }
                        messages.add(ChatLine("Zoom IA", answer))
                        isSending = false
                    }
                }
            ) {
                Text(if (isSending) "Enviando..." else "Enviar")
            }
        }
    }
}

private fun callChatApi(
    endpoint: String,
    model: String,
    apiKey: String,
    prompt: String
): String {
    return try {
        val connection = (URL(endpoint).openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = 30_000
            readTimeout = 60_000
            doOutput = true
            setRequestProperty("Content-Type", "application/json")
            if (apiKey.isNotBlank()) {
                setRequestProperty("Authorization", "Bearer $apiKey")
            }
        }

        val payload = JSONObject().apply {
            put("model", model)
            put(
                "messages",
                JSONArray().put(
                    JSONObject().apply {
                        put("role", "user")
                        put("content", prompt)
                    }
                )
            )
        }

        connection.outputStream.bufferedWriter(Charsets.UTF_8).use { writer ->
            writer.write(payload.toString())
        }

        val code = connection.responseCode
        val stream = if (code in 200..299) connection.inputStream else connection.errorStream
        val body = stream?.bufferedReader(Charsets.UTF_8)?.use { it.readText() }.orEmpty()
        connection.disconnect()

        if (code !in 200..299) {
            return "Erro HTTP $code: ${body.take(500)}"
        }

        val root = JSONObject(body)
        val content = root
            .optJSONArray("choices")
            ?.optJSONObject(0)
            ?.optJSONObject("message")
            ?.optString("content")
            ?.trim()

        if (content.isNullOrBlank()) {
            "A API respondeu, mas não encontrei choices[0].message.content."
        } else {
            content
        }
    } catch (e: Exception) {
        "Falha ao conectar: ${e.message ?: e.javaClass.simpleName}"
    }
}
